### ISSUE TITLE
Lookup for generators in the current location

### ISSUE THREAD

**Issue by ericmorand** [yeoman/yo#545]
## Type of issue: Feature request

I have two problems with yeoman that I think should be discussed and resolved.

The first is that you have to install yeoman globally to have it work. And I just can't ask my developers to install something globally on their computer just to work on a project. A project should be self-contained and usable with nothing else that itself and its dependencies.

The second is that I absolutely don't want to publish my generators on NPM. They are copyright protected by my company, and each one are finely crafted specifically for each project.

Currently, yeoman's approach is based on NPM-published generators and a globally installed CLI tool. Or locally installed generators registered globally with `npm link`.

To cope with my strong principles (the main being: a good project must be self-contained), I had to find a way to use yeoman locally and have it lookup for generators in the local folder.

It actually was quite easy to do. I wrote a yo.js script that can be exceuted by nodejs and...voila! This script is located at the root of each project, yeoman is added as a dev dependency of the project and the generators are maintained locally inside the project.

The thing is, the script is so ugly and stupid that I feel bad for using it. Take a look at this:

```
let path = require('path');
let resolver = require('yeoman-environment/lib/resolver');

resolver.getNpmPaths = function() {
  return [path.join(__dirname, 'lib')];
};

require('yo/lib/cli');
```

As you can see, the script consist in registering the current path (well, `lib` subpath in this case) to the yeoman resolver and calling the cli. But that doesn't feel right to me: replacing a function inside a dependency and calling a cli with a require is not what I consider a good practice.

I would rather write something like this:

```
let yo = require('yo');

yo.registerLocalGenerators(path.join(__dirname, 'lib'));
yo.run();
```

Basically the same thing but safe and clean.

Before trying to add something like this to Yeoman, I'd like to know what the community thinks about it. And ensure if it is not already on the roadmap.

**Comment by mischah** [comment]
mhmm. Not sure if I get your points.

1. You can of course install yo locally just like every other node cli
    - You could even use it without installing at all via `npx`
2. You can use local generators via `yo ./path/to/some/generator`
    - or use npm to install your generator from a private Git endpoint instead of the public registry.

Cheers, Michael

**Comment by ericmorand** [comment]
Hmmm, maybe something is wring with my setup, because after having installed it via `npm install yo --save-dev` I get `yo: command not found` when I execute the command `yo`.

**Comment by mischah** [comment]
> And I didn't know you could execute a generator by its path. That's perfect!

Guess that’s something we should document more prominent. Going to open a separate issue for that.

> Hmmm, maybe something is wrong with my setup, because after having installed it via `npm install yo --save-dev` I get `yo: command not found` when I execute the command `yo`.

Yup. There are three possibilities to use a locally installed CLI in your project:
1. use `npx` 
2. explicitly call the locally installed binary via `node_modules/.bin/yo`
3. setup a npm script by adding the following to the `package.json` of your project
   ```
   "scripts": {
       "yo": "yo --version",
   }
   ```
  Calling `npm run yo` executes the defined command using the locally installed version.

**Comment by ericmorand** [comment]
This is awesome!

I've been using npm for years and I didn't know you could execute locally installed CLI with it. Thanks a lot for your help, it works perfectly. I can know happily delete my script and update my NPM scripts accordingly.

Many thanks.

**Comment by mischah** [comment]
My pleasure 😘

**Comment by SBoudrias** [comment]
Also, with yarn, `yarn xxx` will automatically run any local binaries. No need for a specific script in `package.json`.

**Comment by mischah** [comment]
Oh. Nice, didn’t knew this.
